### PR TITLE
Add semantic versioning

### LINKED ISSUE #321: Versioning of the OpenQASM3 Language Specification
### What is the expected enhancement?

When developing the language specification it is often necessary to break up work into digestible chunks, eg., first update the language specification and then update the grammar. This came up, in particular, to follow on work in #283. It would be desirable to have some sort of versioning strategy for the library so that we could decide when the repository is in a consistent state to release. 

A potential way of doing this would be to a [semantic versioning number](https://semver.org/) for the language release of the form `3.x.x` that is used to tag a Github release and build relevant packages through CI. Within the language itself this could then be specified as 

```qasm 
// First non-comment is a version string
OPENQASM 3.x.x;
```

### PR DESCRIPTION
**Pull request by taalexander**
### Summary
Closes #321. This PR proposes a process for semantic versioning releases of the OpenQASM3 language specification and3 project as discussed in #321 and the TSC meeting.

Moving forward we should:
- [x] Agree on semantic versioning procedure (this PR)
- [x] Determine which version / commit should be cut first. Should `master` become `3.0.0` or `3.1.0`. If it becomes `3.1.0` what commit should we tag as `3.0.0`? A: See Steven's comment below
- [x] Release / tag the current version : A: On hold to June 15th
- [x] Add CI support for building documentation and posting the correct location. A: See #330.
- [x] Add release notes support. A: See #328.

### Details and comments
This PR proposes a development and release cycle for the OpenQASM3 language. 

An example is shown below, see the PR for more details.

#### Example release cycle

For this example assume the current release of OpenQASM is version `3.1.1`. This will correspond to a commit
on `stable/3.1`. The project's development branch reflects the development state of the next release - `3.2.0`
and is referred to by version as `3.2.0-dev`.

To trigger a bugfix release - `3.1.2`:
1. Create a PR into `stable/3.1` with all required changes. These may be backported commits from `3.2.0-dev`.
2. Upon the merger of the PR tag the HEAD of `stable/3.1` with `3.1.2` and push to Github.

To trigger a minor release - `3.2.0`:
1. Create a new stable branch `stable/3.2` using the current development branch as the base branch, eg., `git checkout -b stable/3.2 master`.
2. Push this branch to Github.
3. Tag the branch with `3.2.0` and push to Github.

### CODE REVIEW

**Comment by stevenheidel** [comment]
>  Determine which version / commit should be cut first. Should master become 3.0.0 or 3.1.0. If it becomes 3.1.0 what commit should we tag as 3.0.0?

From our TSC meeting - we will work towards a 3.0 release. We set a "strawman" date of June 15, 2022 for this release.

**Review by levbishop — APPROVED** [review]

**Comment by jwoehr** [comment]
What again do we do with the already extant `3.0` tag?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
+# Contributing
+
+**We appreciate all kinds of help, so thank you!** :clap: :kissing_heart:
+
+You can contribute in many ways to this project.
+
+## Issue reporting
+
+:fire: This is a good point to start, when you find a problem please add it to the [issue tracker](https://github.com/Qiskit/openqasm/issues). Please, use [this template](https://github.com/Qiskit/openqasm/contributing/blob/main/templates/issue.md) to add them.
```
Comment (stevenheidel) [inline CONTRIBUTING.md:9]: We don't have this template and I doubt we need it. Suggest dropping this sentence.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+- Proposal: An idea with the correct form:
+  - Add one issue in the [main repo](https://github.com/Qiskit/openqasm) labeled as `proposal` using [this template](templates/proposal.md).
+- Candidate: During each monthly meeting the assistants select the ones considered more interesting to pass to the next stage. One of the core devs will start commenting on the issue to guide the owner into the next steps, including:
+  - Fork [the main repo](https://github.com/Qiskit/openqasm).
+  - Add the content of the proposal, note that conformance tests are mandatory at this point.
+  - Make a pull request.
+  - The core dev can ask for changes before reaching the next stage.
+- Accepted: When the PR is merged into main.
+
+## Tests
```
Comment (stevenheidel) [inline CONTRIBUTING.md:56]: This section looks copied from qiskit and is not correct for this repo's tests.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+The test runner uses all the circuit files in the [examples](examples) folder. They are run automatically to check they keep passing the parser. It allows dropping more files in those folders, even adding new ones.
+
+- The `invalid` folder includes circuits that should raise a `QasmException`.
+- The rest include valid circuits.
+- Optionally, they can include metadata in the header (inside comments, like [this one](examples/invalid/gate_no_found.qasm)):
+  - name: Descriptive name for the check this example is covering.
+  - section: Link to the related part of the specification.
+
+### Run
```
Comment (stevenheidel) [inline CONTRIBUTING.md:70]: Same here

**Comment by stevenheidel** [comment]
> What again do we do with the already extant `3.0` tag?

Let's delete the tag.

**Comment by stevenheidel** [comment]
@taalexander - I approve of the release process, but looks like the contribution guide was contains some stuff from qiskit that needs to be updated/removed.

**Comment by taalexander** [comment]
> @taalexander - I approve of the release process, but looks like the contribution guide was contains some stuff from qiskit that needs to be updated/removed.

I will remove these however, it should be noted that I just moved `contributing.md` to `CONTRIBUTING.md` for consistency. The commented lines were already in the existing repo. Its just showing all lines changed as Git diffs freak out when you change a moved file.

**Review by stevenheidel — APPROVED** [review]

File: CONTRIBUTING.md (hunk 3 above)
Comment (taalexander) [reply CONTRIBUTING.md]: Fixed in 4e8bb1e.

File: CONTRIBUTING.md (hunk 1 above)
Comment (taalexander) [reply CONTRIBUTING.md]: Fixed in 4e8bb1e.

File: CONTRIBUTING.md (hunk 2 above)
Comment (taalexander) [reply CONTRIBUTING.md]: Fixed in 4e8bb1e.

**Review by jwoehr — APPROVED** [review]
Looks good ... if all @stevenheidel 's concerns have been met, let's merge.

**Review by stevenheidel — APPROVED** [review]

**Pull request merged**